packages/sdk/src/protocols/base_protocol.ts

```diff
@@ -1,82 +1,73 @@
-  ProtocolUseOptions
-} from "@waku/interfaces";
-import { delay, Logger } from "@waku/utils";
+import { IBaseProtocolSDK, ProtocolUseOptions } from "@waku/interfaces";
+import { Logger } from "@waku/utils";
+
+import { PeerManager } from "./peer_manager.js";
 
 interface Options {
   numPeersToUse?: number;
   maintainPeersInterval?: number;
 }
 
-const RENEW_TIME_LOCK_DURATION = 30 * 1000;
 const DEFAULT_NUM_PEERS_TO_USE = 2;
 const DEFAULT_MAINTAIN_PEERS_INTERVAL = 30_000;
 
 export class BaseProtocolSDK implements IBaseProtocolSDK {
-  private healthManager: IHealthManager;
+  private peerManager: PeerManager;
   public readonly numPeersToUse: number;
-  private peers: Peer[] = [];
   private maintainPeersIntervalId: ReturnType<
     typeof window.setInterval
   > | null = null;
   private log: Logger;
 
-  private maintainPeersLock = false;
-  private readonly renewPeersLocker = new RenewPeerLocker(
-    RENEW_TIME_LOCK_DURATION
-  );
-
   public constructor(
     protected core: BaseProtocol,
     protected connectionManager: ConnectionManager,
     options: Options
   ) {
     this.log = new Logger(`sdk:${core.multicodec}`);
 
-    this.healthManager = getHealthManager();
+    this.peerManager = new PeerManager(connectionManager, core, this.log);
 
     this.numPeersToUse = options?.numPeersToUse ?? DEFAULT_NUM_PEERS_TO_USE;
     const maintainPeersInterval =
       options?.maintainPeersInterval ?? DEFAULT_MAINTAIN_PEERS_INTERVAL;
 
+    this.log.info(
+      `Initializing BaseProtocolSDK with numPeersToUse: ${this.numPeersToUse}, maintainPeersInterval: ${maintainPeersInterval}ms`
+    );
+    // void this.setupEventListeners();
     void this.startMaintainPeersInterval(maintainPeersInterval);
   }
 
   public get connectedPeers(): Peer[] {
-    return this.peers;
+    return this.peerManager.getPeers();
   }
 
   /**
    * Disconnects from a peer and tries to find a new one to replace it.
    * @param peerToDisconnect The peer to disconnect from.
    * @returns The new peer that was found and connected to.
    */
-  public async renewPeer(peerToDisconnect: PeerId): Promise<Peer> {
-    this.log.info(`Renewing peer ${peerToDisconnect}`);
+  public async renewPeer(peerToDisconnect: PeerId): Promise<Peer | undefined> {
+    this.log.info(`Attempting to renew peer ${peerToDisconnect}`);
 
-    const peer = (await this.findAndAddPeers(1))[0];
-    if (!peer) {
-      throw Error("Failed to find a new peer to replace the disconnected one.");
+    const newPeer = await this.peerManager.findPeers(1);
+    if (newPeer.length === 0) {
+      this.log.error(
+        "Failed to find a new peer to replace the disconnected one"
+      );
+      return undefined;
     }
 
-    const updatedPeers = this.peers.filter(
-      (peer) => !peer.id.equals(peerToDisconnect)
-    );
-    this.updatePeers(updatedPeers);
-    await this.connectionManager.dropConnection(peerToDisconnect);
+    await this.peerManager.removePeer(peerToDisconnect);
```

Comment:
maybe `.replace` operation should be implemented with a lock inside? this should be enough so we can to avoid usage of extra lib

Reply:
nit: 
of course, but there is no lock between remove and add - tho doesn't seem to be a problem

but from conceptual point - if we use `add/remove` together a lot - makes sense to have replace

Reply:
What we do is:
```ts
await this.removePeer();
await this.addPeer();
```
And we only do it in `BaseProtocol` sequentially. Not required to combine them